src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -239,17 +236,76 @@ def _get_source_file(obj_dir: str, obj_file: str) -> str:
     return src_file
 
 
+def _upload_genomes_to_workspace(
+    gfu: GenomeFileUtil,
+    workspace_id: int,
+    load_id: str,
+    ws_obj_tuples: list[WSObjTuple],
+    data_dir: str,
+) -> list[UploadResult]:
+    """
+    Upload genbank files to the target workspace as Genome in batch. The bulk method fails
+    and an error will be thrown if any of the genome files in batch fails to upload.
+    The order of elements in the returned list corresponds to the order of `ws_obj_tuples`.
+    """
+
+    inputs = [
+        {
+              # GFU has the capability to discern and retrieve the accurate GenBank file from the directory
+            "file": {'path': obj_tuple.container_internal_file_dir},
+            "genome_name": obj_tuple.obj_name,
+            "metadata": {"load_id": load_id},
+        }
+        for obj_tuple in ws_obj_tuples
+    ]
+
+    results = gfu.genbanks_to_genomes(
+        {"workspace_id": workspace_id, "inputs": inputs}
+    )["results"]
+
+    upload_results = []
+    for result_dict, obj_tuple in zip(results, ws_obj_tuples):
+
+        genome_obj_info = result_dict["genome_info"]
+        assembly_obj_info = result_dict["assembly_info"]
+
+        genome_tuple = obj_tuple
+
+        # copy assembly file to the source directory
+        assembly_path = Path(result_dict["assembly_path"])
+        # remove prefix of the assembly_path
+        assembly_path = assembly_path.relative_to(_JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER)
+        assembly_path_local = Path(data_dir) / assembly_path
+        if not os.path.exists(assembly_path_local):
+            raise ValueError(f"Assembly file {assembly_path_local} does not exist")
+
+        source_dir = Path(genome_tuple.host_file_dir)
+        loader_helper.create_hardlink_between_files(source_dir / assembly_path.name, assembly_path_local)
```

Comment:
each genome has its own data_dir. It's putting the associated assembly file into the associated genome data_dir.

Reply:
assembly_path_local is `data_dir/assembly_path` where `data_dir` is the same for all genomes. `assembly_path` is returned from assembly util with `/kb/module/work/tmp` trimmed off the front. Does that mean we're relying on assembly util to match up with our directory structure in the collections source dir?

Reply:
> Does that mean we're relying on assembly util to match up with our directory structure in the collections source dir?

hmmm. I am not following this. This is to copy/link FASTA file returned from AU to `sourcedata/NCBI/None/<genome_id>` directory. We are not relying on AU to match up collection directory structure.

Reply:
Actually this is even more confusing than I thought. The code creates a hardlink at source_dir / assembly_path.name pointing to assembly_path_local. But IIUC the former should already exist since that's where GFU places the file.

Reply:
no. GFU places the FASTA file at the tmp job_dir which is the `assembly_path_local`. We want to link this file to NCBI source dir (sourcedata/NCBI/None). We have only downloaded the genbank file at NCIB source dir.

Reply:
maybe we don't wan to link the FASTA file to NCBI source dir? Just link it at WS source dir? So that,
in sourcedata/NCBI, we have genbank file and uploaded.yaml
and in sourcedata/WS, we have FASTA file and metadata file

Reply:
no it actually doesn't make sense. Because we have `assembly_filename` in the uploaded.yaml file. I am guess we assume this file does exist in the NCBI source dir.

Reply:
currently I am leaving this FASTA file in both NCBI and WS sourcedata dir. I think that makes the most sense. 
so in sourcedata/NCBI, we have genbank file, FASTA file and uploaded.yaml
and in sourcedata/WS, we have FASTA file and metadata file

Reply:
And IIUC the fasta files are hardlinked, so there's only one physical copy of the file?

Reply:
yea. one physical file in the job_dir

Reply:
`<pedandic>`
ACKSHULLALY, the job_dir file is just an entry in the file table that points to an inode, and the new hardlink just points to the same inode. 
`</pedantic>`

Reply:
It turns out `host_file_dir` is *not* the equivalent of `container_internal_file_dir` on the host - it's actually the `collectionssource/NONE/<collection>/<source_ver>/<genome_id>` dir. Can you check that's consistent across the code base, and if so, rename? That confused the heck out of me

Reply:
👍

Reply:
:+1: